### PR TITLE
Suppress wallet exception

### PR DESCRIPTION
**Pull request by quantumagi**
See BUG 3002.

https://dev.azure.com/Stratisplatformuk/StratisBitcoinFullNode/_workitems/edit/3002/

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Wallet/Wallet.cs [hunk 1]
```diff
@@ -98,11 +98,15 @@ public IEnumerable<HdAccount> GetAccountsByCoinType(CoinType coinType, Func<HdAc
...
         /// <param name="coinType">The type of the coin this account is for.</param>
+        /// <param name="missingThrowsError">Set to <c>true</c> to throw an error if the account is missing and <c>false</c> to return <c>null</c>.</param>
         /// <returns>The requested account.</returns>
-        public HdAccount GetAccountByCoinType(string accountName, CoinType coinType)
+        public HdAccount GetAccountByCoinType(string accountName, CoinType coinType, bool missingThrowsError = true)
         {
             AccountRoot accountRoot = this.AccountsRoot.SingleOrDefault(a => a.CoinType == coinType);
-            return accountRoot?.GetAccountByName(accountName);
+            if (accountRoot == null && missingThrowsError)
+                    throw new WalletException($"No account with the name '{accountName}' could be found.");
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Wallet/Wallet.cs]: 1 redundant tab (4 spaces)

File: src/Stratis.Bitcoin.Features.Wallet/Wallet.cs [hunk 2]
```diff
@@ -98,11 +98,15 @@ public IEnumerable<HdAccount> GetAccountsByCoinType(CoinType coinType, Func<HdAc
         /// </summary>
         /// <param name="accountName">The name of the account to retrieve.</param>
         /// <param name="coinType">The type of the coin this account is for.</param>
+        /// <param name="missingThrowsError">Set to <c>true</c> to throw an error if the account is missing and <c>false</c> to return <c>null</c>.</param>
         /// <returns>The requested account.</returns>
-        public HdAccount GetAccountByCoinType(string accountName, CoinType coinType)
+        public HdAccount GetAccountByCoinType(string accountName, CoinType coinType, bool missingThrowsError = true)
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Wallet/Wallet.cs]: not sure if `bool missingThrowsError` is a good approach. Probably better would be to go with single way to handle missing account and enforce users of this method to handle the result as they please

[the remark this answers is not in the document]
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Wallet/Wallet.cs]: Changed to return null.

File: src/Stratis.Bitcoin.Features.Wallet/Wallet.cs (hunk 1 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Wallet/Wallet.cs]: Fixed.

File: src/Stratis.Bitcoin.Features.Wallet/Wallet.cs (hunk 2 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Wallet/Wallet.cs]: Done.

File: src/Stratis.Bitcoin.Features.SmartContracts/Wallet/SmartContractTransactionService.cs [hunk 3]
```diff
@@ -49,55 +47,51 @@ public class SmartContractTransactionService : ISmartContractTransactionService
 
         public BuildCallContractTransactionResponse BuildCallTx(BuildCallContractTransactionRequest request)
         {
-            AddressBalance addressBalance = this.walletManager.GetAddressBalance(request.Sender);
-            if (addressBalance.AmountConfirmed == 0 && addressBalance.AmountUnconfirmed == 0)
-                return BuildCallContractTransactionResponse.Failed(SenderNoBalanceError);
+            try
```
Comment (quantumagi) [inline src/Stratis.Bitcoin.Features.SmartContracts/Wallet/SmartContractTransactionService.cs]: Changed to use a single try / catch block.

File: src/Stratis.Bitcoin.Features.Wallet.Tests/AccountRootTest.cs [hunk 4]
```diff
@@ -59,7 +59,7 @@ public void GetAccountByNameWithNonMatchingNameThrowsException()
         {
             AccountRoot accountRoot = CreateAccountRootWithHdAccountHavingAddresses("Test", CoinType.Stratis);
 
-            Assert.Throws<WalletException>(() => { accountRoot.GetAccountByName("test"); });
+            Assert.Null(accountRoot.GetAccountByName("test"));
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Wallet.Tests/AccountRootTest.cs]: Please change the name of the test.

Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Wallet.Tests/AccountRootTest.cs]: Done.

File: src/Stratis.Bitcoin.Features.SmartContracts/Wallet/SmartContractTransactionService.cs (hunk 3 above)
Comment (bokobza) [reply src/Stratis.Bitcoin.Features.SmartContracts/Wallet/SmartContractTransactionService.cs]: You've included a bunch of statements that weren't inside the try before.

Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.SmartContracts/Wallet/SmartContractTransactionService.cs]: Yes, the `try` statement was moved up. Also we only have a single `try` / `catch` now. The reasoning is that we probably need a `BuildCallContractTransactionResponse` regardless of the type of error that occurs. Only selectively handling errors may not be that good an idea.

Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.SmartContracts/Wallet/SmartContractTransactionService.cs]: Note that `BuildCallTx` is called from the `Call` http post method that does not have its own exception handling...

Comment (bokobza) [reply src/Stratis.Bitcoin.Features.SmartContracts/Wallet/SmartContractTransactionService.cs]: so you've change the functionality here a bit.
Can you please just change the account name exception thing?

**Comment by bokobza** [comment]
I'll merge when the builds are green.

**Pull request merged**